File: app/src/main/java/com/arejas/dashboardofthings/presentation/ui/helpers/HistoryChartHelper.java
package com.arejas.dashboardofthings.presentation.ui.helpers;

import com.arejas.dashboardofthings.DotApplication;
import com.arejas.dashboardofthings.R;
import com.arejas.dashboardofthings.utils.Enumerators;
import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.formatter.DefaultValueFormatter;
import com.github.mikephil.charting.formatter.ValueFormatter;

import java.text.SimpleDateFormat;
import java.util.Date;

public class HistoryChartHelper {

    public static final int SPINNER_HISTORY_LASTVAL = 0;
    public static final int SPINNER_HISTORY_LASTDAY = 1;
    public static final int SPINNER_HISTORY_LASTWEEK = 2;
    public static final int SPINNER_HISTORY_LASTMONTH = 3;
    public static final int SPINNER_HISTORY_LASTYEAR = 4;

    public static final long ONE_HOUR_MILLIS = 3600*1000;

    public static ValueFormatter getValueFormatterForTimeAxis(int spinnerHistorySelection,
                                                              Date firstDate, Date lastDate) {
        return new ValueFormatterXAxisDate(spinnerHistorySelection, firstDate, lastDate);
    }

    public static ValueFormatter getValueFormatterForDataAxis(Enumerators.DataType dataType) {
        switch (dataType) {
            case BOOLEAN:
                return new ValueFormatter() {
                    @Override
                    public String getAxisLabel(float value, AxisBase axis) {
                        if (value == 1.0f) {
                            return  DotApplication.getContext().getString(R.string.boolean_active);
                        } else {
                            return DotApplication.getContext().getString(R.string.boolean_not_active);
                        }
                    }
                };
            case INTEGER:
                return new DefaultValueFormatter(0);
            default:
                return new DefaultValueFormatter(1);
        }
    }

    static class ValueFormatterXAxisDate extends ValueFormatter {

        final int spinnerHistoryValue;
        final Date dateMin;
        final Date dateMax;

        public ValueFormatterXAxisDate(int spinnerHistoryValue, Date dateMin, Date dateMax) {
            this.spinnerHistoryValue = spinnerHistoryValue;
            this.dateMin = dateMin;
            this.dateMax = dateMax;
        }

        @Override
        public String getAxisLabel(float value, AxisBase axis) {
            try {
                Date date = new Date(Float.valueOf(value).longValue() + dateMin.getTime());
                switch (spinnerHistoryValue) {
                    case SPINNER_HISTORY_LASTDAY:
                        return (new SimpleDateFormat("HH:00")).format(date);
                    case SPINNER_HISTORY_LASTWEEK:
                        return (new SimpleDateFormat("E")).format(date);
                    case SPINNER_HISTORY_LASTMONTH:
                        return (new SimpleDateFormat("dd/MM")).format(date);
                    case SPINNER_HISTORY_LASTYEAR:
                        return (new SimpleDateFormat("MMM yy")).format(date);
                    default:
                        long difference = Math.abs(dateMax.getTime() - dateMin.getTime());
                        if (difference < ONE_HOUR_MILLIS) {
                            return (new SimpleDateFormat("HH:mm:ss")).format(date);
                        }
                }
                return (new SimpleDateFormat("dd/MM HH:mm:ss")).format(date);
            } catch (Exception e) {
                return "NNONONONO";
            }
        }
    }

}

File: app/src/main/java/com/arejas/dashboardofthings/presentation/ui/helpers/HistoryChartHelper.kt
package com.arejas.dashboardofthings.presentation.ui.helpers

import com.arejas.dashboardofthings.DotApplication
import com.arejas.dashboardofthings.R
import com.arejas.dashboardofthings.utils.Enumerators
import com.github.mikephil.charting.components.AxisBase
import com.github.mikephil.charting.formatter.DefaultValueFormatter
import com.github.mikephil.charting.formatter.ValueFormatter

import java.text.SimpleDateFormat
import java.util.Date

object HistoryChartHelper {

    val SPINNER_HISTORY_LASTVAL = 0
    val SPINNER_HISTORY_LASTDAY = 1
    val SPINNER_HISTORY_LASTWEEK = 2
    val SPINNER_HISTORY_LASTMONTH = 3
    val SPINNER_HISTORY_LASTYEAR = 4

    val ONE_HOUR_MILLIS = (3600 * 1000).toLong()

    fun getValueFormatterForTimeAxis(
        spinnerHistorySelection: Int,
        firstDate: Date, lastDate: Date
    ): ValueFormatter {
        return ValueFormatterXAxisDate(spinnerHistorySelection, firstDate, lastDate)
    }

    fun getValueFormatterForDataAxis(dataType: Enumerators.DataType): ValueFormatter {
        when (dataType) {
            Enumerators.DataType.BOOLEAN -> return object : ValueFormatter() {
                override fun getAxisLabel(value: Float, axis: AxisBase?): String {
                    return if (value == 1.0f) {
                        DotApplication.context.getString(R.string.boolean_active)
                    } else {
                        DotApplication.context.getString(R.string.boolean_not_active)
                    }
                }
            }
            Enumerators.DataType.INTEGER -> return DefaultValueFormatter(0)
            else -> return DefaultValueFormatter(1)
        }
    }

    internal class ValueFormatterXAxisDate(
        val spinnerHistoryValue: Int,
        val dateMin: Date,
        val dateMax: Date
    ) : ValueFormatter() {

        override fun getAxisLabel(value: Float, axis: AxisBase?): String {
            try {
                val date = Date(java.lang.Float.valueOf(value).toLong() + dateMin.time)
                when (spinnerHistoryValue) {
                    SPINNER_HISTORY_LASTDAY -> return SimpleDateFormat("HH:00").format(date)
                    SPINNER_HISTORY_LASTWEEK -> return SimpleDateFormat("E").format(date)
                    SPINNER_HISTORY_LASTMONTH -> return SimpleDateFormat("dd/MM").format(date)
                    SPINNER_HISTORY_LASTYEAR -> return SimpleDateFormat("MMM yy").format(date)
                    else -> {
                        val difference = Math.abs(dateMax.time - dateMin.time)
                        if (difference < ONE_HOUR_MILLIS) {
                            return SimpleDateFormat("HH:mm:ss").format(date)
                        }
                    }
                }
                return SimpleDateFormat("dd/MM HH:mm:ss").format(date)
            } catch (e: Exception) {
                return "NNONONONO"
            }

        }
    }

}
